File: KargoPaylasimSwiftUI/KargoPaylasimSwiftUI/service/ApiClient.swift
import Foundation

final class APIClient {
    let tokenStore: TokenStore

    init(tokenStore: TokenStore) {
        self.tokenStore = tokenStore
    }

    func get<T: Decodable>(_ ep: Endpoint, query: [String:String] = [:], as: T.Type) async throws -> T {
        var comp = URLComponents(url: ep.url, resolvingAgainstBaseURL: false)!
        if !query.isEmpty {
            comp.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        guard let url = comp.url else { throw APIError.badURL }

        var req = URLRequest(url: url)
        req.httpMethod = "GET"
        if let t = tokenStore.token { req.setValue(t, forHTTPHeaderField: "X-Auth-Token") }

        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw APIError.badResponse }
        if http.statusCode == 401 { throw APIError.unauthorized }

        do { return try JSONDecoder().decode(T.self, from: data) }
        catch {
            let raw = String(data: data, encoding: .utf8) ?? ""
            throw APIError.server("Decode error. Raw:\n\(raw)")
        }
    }

    func postJSON<B: Encodable, T: Decodable>(_ ep: Endpoint, body: B, as: T.Type) async throws -> T {
        var req = URLRequest(url: ep.url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let t = tokenStore.token { req.setValue(t, forHTTPHeaderField: "X-Auth-Token") }
        req.httpBody = try JSONEncoder().encode(body)

        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw APIError.badResponse }
        if http.statusCode == 401 { throw APIError.unauthorized }

        do { return try JSONDecoder().decode(T.self, from: data) }
        catch {
            let raw = String(data: data, encoding: .utf8) ?? ""
            throw APIError.server("Decode error. Raw:\n\(raw)")
        }
    }
}

File: flutter_kargo_paylasim/lib/service/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../storage/token_store.dart';

class ApiError implements Exception {
  final String message;
  ApiError(this.message);
  @override
  String toString() => message;
}

class ApiClient {
  final TokenStore tokenStore;
  ApiClient(this.tokenStore);

  Future<Map<String, String>> _headers({bool jsonBody = true}) async {
    final h = <String, String>{};
    if (jsonBody) h["Content-Type"] = "application/json";
    final token = await tokenStore.getToken();
    if (token != null && token.isNotEmpty) {
      h["X-Auth-Token"] = token;
    }
    return h;
  }

  Future<Map<String, dynamic>> postJson(String url, Map<String, dynamic> body) async {
    final res = await http.post(
      Uri.parse(url),
      headers: await _headers(),
      body: jsonEncode(body),
    );
    return _decodeOrThrow(res);
  }

  Future<Map<String, dynamic>> getJson(String url, {Map<String, String>? query}) async {
    final uri = Uri.parse(url).replace(queryParameters: query);
    final res = await http.get(uri, headers: await _headers(jsonBody: false));
    return _decodeOrThrow(res);
  }

  Map<String, dynamic> _decodeOrThrow(http.Response res) {
    final raw = res.body;
    try {
      final j = jsonDecode(raw);
      if (j is Map<String, dynamic>) return j;
      throw ApiError("Invalid JSON");
    } catch (_) {
      throw ApiError("Decode error. Raw:\n$raw");
    }
  }
}
